### PR TITLE
Kernel: Limit free space between randomized memory allocations

### PR DESCRIPTION
**Pull request by timschumi**
This is an attempt to salvage more memory from our address space by limiting how far away from existing mappings new randomized mappings can be created.

It works by allowing completely randomized placements until a certain threshold (in this case, 100 allocated regions) to allow for more possible placements once the limiter is active. After that, the distance between the attempted mapping and the nearest existing mapping is checked and, if the distance is bigger than a certain threshold, rejected.

Note that, as a relative newcomer on the topic, I just picked the thresholds basically arbitrary, as I can't make any educated guesses on what thresholds would be most "efficient". However, I do plan to go back to my original test case and see how much space I can reasonably let it waste before running out of memory again. Also, I haven't yet put in support for placing allocations just _before_ existing mappings, so those will be erroneously rejected in the initial version of the PR.

This works around an issue where we would run out of memory while linking a bunch of small object files against a large library using Clang, as it would map the (rather small) object files with multiple megabytes of free space in between, and end up not being able to fit the 60MB behemoth that is `libLLVM.so`.

### CODE REVIEW

**Comment by IdanHo** [comment]
Can we do this only on i686? x86_64 has no issue with regards to available vaddr space.

**Comment by timschumi** [comment]
I now updated the limiter to only apply to i686, and to consider memory regions just before regions that are already allocated. I also tuned the limits to be way less restrictive, which should lessen the performance impact on anything that doesn't allocate space for hundreds of files (if there is any better metric for that than "number of allocated regions", I'm all ears).

File: Kernel/Memory/RegionTree.cpp [hunk 1]
```diff
@@ -121,6 +121,23 @@ ErrorOr<VirtualRange> RegionTree::allocate_range_randomized(size_t size, size_t
         if (!m_total_range.contains(random_address, size))
             continue;
 
+#if ARCH(I386)
```
Comment (awesomekling) [inline Kernel/Memory/RegionTree.cpp:124]: Please add a small comment here describing what this is about. :^)

File: Kernel/Memory/RegionTree.cpp [hunk 2]
```diff
@@ -121,6 +121,23 @@ ErrorOr<VirtualRange> RegionTree::allocate_range_randomized(size_t size, size_t
...
+        VirtualAddress random_address_end { random_address.get() + size };
+        constexpr size_t limiter_minimum_allocations = 200;
+        constexpr size_t limiter_maximum_space = 1 * MiB;
```
Comment (awesomekling) [inline Kernel/Memory/RegionTree.cpp:127]: What's a "limiter" in this context?

**Review by awesomekling — CHANGES_REQUESTED** [review]

File: Kernel/Memory/RegionTree.cpp (hunk 2 above)
Comment (timschumi) [reply Kernel/Memory/RegionTree.cpp]: Nothing in particular. I renamed the variables to hopefully make it more clear.

File: Kernel/Memory/RegionTree.cpp (hunk 1 above)
Comment (timschumi) [reply Kernel/Memory/RegionTree.cpp:124]: Done (?)

**Pull request merged**